# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1759259942 -7200
# Parent  29b0701c894eeebaa3a552f24db452e61235f721
Bug 637830 - introduce pref mail.notification.show_on_app_screen.
* * *
NNN28 - Fix flashing new mail alert window.

diff --git a/mailnews/base/src/MailNotificationManager.sys.mjs b/mailnews/base/src/MailNotificationManager.sys.mjs
--- a/mailnews/base/src/MailNotificationManager.sys.mjs
+++ b/mailnews/base/src/MailNotificationManager.sys.mjs
@@ -480,15 +480,51 @@ export class MailNotificationManager {
     args.appendElement({
       wrappedJSObject: newMsgKeys,
     });
     args.appendElement(this);
-    Services.ww.openWindow(
+    const win = Services.ww.openWindow(
       null,
       "chrome://messenger/content/newmailalert.xhtml",
       "_blank",
       "chrome,dialog,titlebar=no,alert=yes",
       args
     );
+    // Move window to its final position, so in case it
+    // flashes, it will at least "flash in place".
+    let screenToUse = win.screen;
+    let scaleX = 0;
+    let scaleY = 0;
+    if (
+      Services.prefs.getBoolPref("mail.notification.show_on_app_screen", false)
+    ) {
+      const appWindow = Services.wm.getMostRecentWindow("mail:3pane");
+      if (
+        appWindow &&
+        (appWindow.screenLeft < 0 || appWindow.screenLeft > win.screen.width)
+      ) {
+        // Application screen to the left or the right of the primary screen.
+        screenToUse = appWindow.screen;
+
+        // If were are on the right screen, and the alert was created on the
+        // left screen, we need to scale compute everything based on the left screen.
+        if (appWindow.screenLeft > win.screen.width) {
+          scaleX = win.screen.width / screenToUse.width;
+          scaleY = win.screen.height / screenToUse.height;
+        }
+      }
+    }
+    let x;
+    let y;
+    // Hack: Assume minimum standard size of 560 x 80.
+    if (scaleX) {
+      x = win.screen.width + screenToUse.width * scaleX - 560 - 10 * scaleX;
+      y = screenToUse.availHeight * scaleY - 80 - 10 * scaleY;
+    } else {
+      x = screenToUse.availLeft + screenToUse.availWidth - 560 - 10;
+      y = screenToUse.availTop + screenToUse.availHeight - 80 - 10;
+    }
+    win.moveTo(x, y);
+
     this._customizedAlertShown = true;
     this._saveNotificationTime(folder, newMsgKeys);
   }
 
diff --git a/mailnews/mailnews.js b/mailnews/mailnews.js
--- a/mailnews/mailnews.js
+++ b/mailnews/mailnews.js
@@ -1028,8 +1028,9 @@ pref("ldap_2.servers.osx.dirType", 3);
 pref("mail.notification.sound", "");
 #endif
 pref("mail.notification.count.inbox_only", true);
 pref("mail.notification.loglevel", "Warn");
+pref("mail.notification.show_on_app_screen", false);
 
 // For the Empty Junk/Trash confirmation dialogs.
 pref("mailnews.emptyJunk.dontAskAgain", false);
 pref("mailnews.emptyTrash.dontAskAgain", false);
